### PR TITLE
Datamovement task for each task and matmul experiment setup

### PR DESCRIPTION
**Pull request by nicelhc13**
This PR updates data movement tasks which were one for each computation task to be one for each computation task's OPERANDS. Expected benefits from these updates are that each data movement task could have different dependencies and could be done earlier than the corresponding parent task (the computation task).

This PR touches data flow graph to support iterator iterating all three data operand lists.

This PR also touches PArray to provide stable data movements from GPU to other GPUs.

Lastly, this PR adds new matmul using explicit data movements as a comparator to matmul_automove.
Both applications are modified to start from CPU memory and give explicit dependencies.
One thing that I failed is to copy back to result (C blocks) to CPU memory since I don't know how to do this by PArray on nested tasks. (Just auto_move() after the inner task didn't work.)
But for performance comparisons, it is not necessary. They can be ignored for now.

### CODE REVIEW

**Comment by dialecticDolt** [comment]
Overall LGTM! 

Only other comment is that some of the iterator use might be hard to port to Cython if it ever needs to be.

**Comment by dialecticDolt** [comment]
Matmul looks good for comparison between the two. I have no idea what you're seeing such a large speedup, that seems weird to me. 

When MSI is up, it's going to need a 2D partitioning on C.

**Comment by nicelhc13** [comment]
@dialecticDolt  regarding performance improvement, only thing that I could think of is data movement task. The original tasks move tasks when the computation tasks start. The new tasks spawn child tasks which move data, and prepare them in advance. For example, let's consider this situation:
A: touches arrays: a, b, and c.
B: touches array b. (dependency: A)
C: touches arrays: b and c (dependency: A and B)

In this case, C creates data movement tasks for b and c, and it would look like this:
C-b: (dependency: A, B)
C-c: (dependency: A)

Therefore, the new one can move array c in advance before B is done, but after A is done.

I think that makes some improvements. 
No MSI is required.

**Review by dialecticDolt — APPROVED** [review]

File: parla/task_runtime.py
Comment (nicelhc13) [reply parla/task_runtime.py:1182]: I think it is necessary. Yes, the data movement tasks should be done before the computation tasks start and after the dependencies start. But the point that I want to make here is that the data movements do not need to wait for all the original tasks dependencies. Each of data movement task has individual dependency. For example, in the above case, C-c does not need to wait for B. C-b needs to wait for both A and B. If C-c just moves data after B is completed, then it would be slower if C-c array is much bigger than C-b.

File: examples/matmul_automove.py
Comment (nicelhc13) [reply examples/matmul_automove.py:72]: We have no clear plan for now. That's what Yineng and I talked about through Slack. But the values should be back to outer array (original array) because other tasks may need it. For example, our blocked cholesky has a bunch of data dependencies among tasks. If we cannot push back the result to the outside, we cannot make cholesky work. I will raise this issue to Friday meeting.

Comment (dialecticDolt) [reply examples/matmul_automove.py:72]: Yeah.. this is tricky. We should definitely gather opinions from Mattan/George/Chris/etc about how this should behave.

**Review by sestephens73 — APPROVED** [review]
LGTM. One thing we need to get around to eventually is cleanly dividing the TaskBase, Task, ComputationTask, and DataMovementTask classes up - it's a little messy right now. But it looks good and you showed that it works today!

File: parla/task_runtime.py
Comment (yinengy) [reply parla/task_runtime.py]: Is current data movement tasks run sync or asynchronously or doesn't matter? Since the current PArray move data without using Stream, cupy.copy() will sync. Does this matter or not?

File: parla/task_runtime.py
Comment (yinengy) [inline parla/task_runtime.py:1267]: What is the use of chain here if you already add an iterator to dataflow?

**Review by yinengy — APPROVED** [review]

File: parla/parray/core.py
Comment (nicelhc13) [reply parla/parray/core.py:63]: I see. It will be merged after your PR is merged. Do you agree with the below update calling copy_from_device_async and do your PR reflect this? The `asarray()` was problematic. As I reported before, it didn't work when I used on scheduler.

File: parla/task_runtime.py
Comment (nicelhc13) [reply parla/task_runtime.py]: Among tasks, they are asynchronous. My first hand thought is that it should not be matter since parallel data movement task moves different data block. Also, since each datamovement task already runs on its own stream, there is no need to use separate stream on PArray. One concern is not your code, but CuPy copy. As I reported before, and left on your PR, it could use its own stream, which is buggy for our Parla. But we'll see whether it works or not.

**Comment by nicelhc13** [comment]
> 

Yes. we should clear class organizations. I am not familiar with Python inheritances and need more studies. Maybe expert folks could help this. Thanks!

File: parla/parray/core.py
Comment (nicelhc13) [reply parla/parray/core.py:63]: Yineng, let me merge this to dev-scheduler branch first. Let me handle your suggestion when we merge this onto dev branch (after your branch is done)

Comment (yinengy) [reply parla/parray/core.py:63]: I think my branch is already done? I am working on find grained movement now.

Comment (nicelhc13) [reply parla/parray/core.py:63]: I think that is also an important choice. Should PArray be thread-safe? Or do the upper layers who will use PArray support thread-safe? I prefer the latter one. I am not for sure if this discussion is about threading. Before doing that, could you discuss with other people first?

**Review by yinengy — APPROVED** [review]

**Comment by nicelhc13** [comment]
> And in regard of copy_from_device_async, my current implementation actually doesn't use asarray. Instead, all data operation is done by cupy.copy() and cupy.array. So, is it still an issue now?

Hm.. I don't know honestly. I don't think so. I will use yours when I merge the new PArray.
(In my opinion, `asarray()` also should work. But it didn't work. I cannot understand why, and so I just tried to use the workaround version which we have used. I think copy() of your PR should work. But.. if that error happened again, let's use this. So let me check this after we merge everything)

**Pull request merged**
